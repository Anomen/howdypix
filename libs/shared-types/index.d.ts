export type MessageProcess = {
  root: string;
  path: string;
};
export type ExifData = {
  ISO?: number;
  make?: string;
  model?: string;
  createDate?: string;
};
