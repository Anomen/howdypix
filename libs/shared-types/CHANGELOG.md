# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- Simplify types for RabbitMQ messages ([#132](https://github.com/vickev/howdypix/pull/132))

## [0.1.0] - 2020-04-09

### Changed

- Version bump because lerna

## [0.0.2] - 2020-03-07

### Added

- Version bump because lerna


