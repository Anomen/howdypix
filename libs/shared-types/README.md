# @howdypix/shared-types

This package exposes common typescript types to be consumed by the HowdyPix applications. For example, it might be the data types sent to RabbitMQ, and consumed on the other side, adding the same type check on both client and server side.
