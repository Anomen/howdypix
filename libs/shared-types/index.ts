export type MessageProcess = {
  root: string;
  path: string;
};

export type ExifData = {
  ISO?: number;
  make?: string;
  model?: string;
  createDate?: string;
};

export type ProcessData = {
  exif: ExifData;
  thumbnails: string[];
};

export enum QueueName {
  TO_PROCESS = "toProcess",
  PROCESSED = "processed"
}
