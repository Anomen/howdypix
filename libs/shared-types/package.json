{
  "name": "@howdypix/shared-types",
  "version": "0.1.0",
  "license": "MIT",
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "scripts": {
    "dev": "tsc --watch",
    "build": "rm -rf build && tsc",
    "lint": "eslint *.{ts,tsx}",
    "typecheck": "tsc --noEmit"
  },
  "devDependencies": {
    "@howdypix/eslint-config": "^0.1.0",
    "@howdypix/typescript-config": "^0.1.0"
  },
  "dependencies": {}
}
