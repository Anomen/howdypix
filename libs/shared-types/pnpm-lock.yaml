dependencies:
  typescript: registry.npmjs.org/typescript/3.6.3
devDependencies:
  '@howdypix/eslint-config': 'link:../eslint-config'
  '@howdypix/typescript-config': 'link:../typescript-config'
lockfileVersion: 5.1
packages:
  registry.npmjs.org/typescript/3.6.3:
    dev: false
    engines:
      node: '>=4.2.0'
    hasBin: true
    name: typescript
    resolution:
      integrity: sha512-N7bceJL1CtRQ2RiG0AQME13ksR7DiuQh/QehubYcghzv20tnh+MQnQIuJddTmsbqYj+dztchykemz0zFzlvdQw==
      registry: 'https://registry.npmjs.org/'
      tarball: 'https://registry.npmjs.org/typescript/-/typescript-3.6.3.tgz'
    version: 3.6.3
specifiers:
  '@howdypix/eslint-config': ^0.1.0
  '@howdypix/typescript-config': ^0.1.0
  typescript: ^3.6.3
