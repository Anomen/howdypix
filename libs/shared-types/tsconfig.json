{
  "extends": "@howdypix/typescript-config/base",
  "exclude": ["node_modules", "**/*.test.ts"],
  "include": ["*.ts"],
  "compilerOptions": {
    "declaration": true,
    "declarationDir": "./build",
    "outDir": "./build"
  }
}
