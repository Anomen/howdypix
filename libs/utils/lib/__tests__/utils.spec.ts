import debug from "debug";
import { omit } from "lodash";
import { HFile, QueueName } from "@howdypix/shared-types";
import { Channel } from "amqplib";
import {
  appDebug,
  appError,
  appInfo,
  appWarning,
  assertQueue,
  consume,
  generateFileUrl,
  generateThumbnailPaths,
  generateThumbnailUrls,
  isHowdypixPath,
  libDebug,
  removeEmptyValues,
  sendToQueue,
  sortJsonStringify,
  wait,
  sortJson,
  parentDir,
  isSupportedMime,
} from "../utils";

jest.mock("debug", () => ({
  __esModule: true,
  default: jest.fn(),
}));

const baseHFile: HFile = {
  name: "name",
  source: "source",
  file: "file",
  dir: "dir",
};

describe("utils::", () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetAllMocks();

    ((debug as unknown) as jest.Mock).mockImplementation(() => (): void => {});
  });

  describe("wait", () => {
    // eslint-disable-next-line jest/no-test-callback
    it("should call setTimeout", async (done) => {
      // Initialize
      jest.useFakeTimers();
      wait(1).then(() => {
        // test
        expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000);
        done();
      });

      // Run
      jest.runAllTimers();
    });
  });

  describe("generateThumbnailPaths", () => {
    it("should generate the right paths", () => {
      expect(
        generateThumbnailPaths("/thumbnails", baseHFile)
      ).toMatchSnapshot();
    });

    it("should generate the right paths with missing dir", () => {
      expect(
        generateThumbnailPaths("/thumbnails", omit(baseHFile, ["dir"]))
      ).toMatchSnapshot();
    });

    it("should generate the right paths with missing file", () => {
      expect(() =>
        generateThumbnailPaths("/thumbnails", omit(baseHFile, ["file"]))
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe("generateThumbnailUrls", () => {
    it("should generate the right URLs", () => {
      expect(
        generateThumbnailUrls("http://toto.com", baseHFile)
      ).toMatchSnapshot();

      expect(
        generateThumbnailUrls("http://toto.com/", baseHFile)
      ).toMatchSnapshot();
    });

    it("should generate the right paths with missing dir", () => {
      expect(
        generateThumbnailUrls("http://toto.com", omit(baseHFile, ["dir"]))
      ).toMatchSnapshot();
    });

    it("should generate the right paths with missing file", () => {
      expect(() =>
        generateThumbnailUrls("http://toto.com", omit(baseHFile, ["file"]))
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe("generateFileUrl", () => {
    it("should generate the right URLs", () => {
      expect(generateFileUrl("http://toto.com", baseHFile)).toMatchSnapshot();

      expect(generateFileUrl("http://toto.com/", baseHFile)).toMatchSnapshot();
    });

    it("should generate the right paths with missing dir", () => {
      expect(
        generateFileUrl("http://toto.com", omit(baseHFile, ["dir"]))
      ).toMatchSnapshot();
    });

    it("should generate the right paths with missing file", () => {
      expect(() =>
        generateFileUrl("http://toto.com", omit(baseHFile, ["file"]))
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe("isHowdypixPath", () => {
    it("should return true if it's an howdypath", () => {
      expect(isHowdypixPath("/asdasd/.howdypix/asdasd")).toEqual(true);
      expect(isHowdypixPath("/asdasd/.wrong/asdasd")).toEqual(false);
      expect(isHowdypixPath()).toEqual(false);
    });
  });

  describe("debugging functions", () => {
    [appInfo, appWarning, appError, appDebug, libDebug].forEach((fct) => {
      it(`${fct.name} should call debug() with the right parameters`, () => {
        fct("test");
        expect(debug).toHaveBeenCalled();
        expect(
          ((debug as unknown) as jest.Mock).mock.calls[0]
        ).toMatchSnapshot();
      });
    });
  });

  describe("assertQueue", () => {
    it("should call assertQueue from the channel", () => {
      const channel = { assertQueue: jest.fn() };
      assertQueue((channel as unknown) as Channel, QueueName.PROCESSED);
      expect(channel.assertQueue).toHaveBeenCalledWith(QueueName.PROCESSED);
    });
  });

  describe("consume", () => {
    const content = { data: 1 };
    const channel = {
      consume: jest.fn(),
    };

    beforeEach(() => {
      channel.consume.mockImplementation((name, onMessage) => {
        onMessage({ content: JSON.stringify({ content }) });
      });
    });

    it("should consume a message with the options", () => {
      consume(
        (channel as unknown) as Channel,
        QueueName.TO_PROCESS,
        undefined,
        { noAck: true }
      );

      expect(channel.consume).toHaveBeenCalledWith(
        QueueName.TO_PROCESS,
        expect.any(Function),
        { noAck: true }
      );
    });

    it("should call the callback with the result of the consume", () => {
      const callback = jest.fn();
      consume((channel as unknown) as Channel, QueueName.TO_PROCESS, callback);
      expect(callback).toHaveBeenCalled();
      expect(callback.mock.calls[0]).toMatchSnapshot();
    });

    it("should call the callback with null if the message is not correct", () => {
      channel.consume.mockImplementation((name, onMessage) => {
        onMessage();
      });

      const callback = jest.fn();
      consume((channel as unknown) as Channel, QueueName.TO_PROCESS, callback);
      expect(callback).toHaveBeenCalledWith(null);
    });
  });

  describe("sendToQueue", () => {
    it("should call sendToQueue with the right options", () => {
      const channel = { sendToQueue: jest.fn() };
      sendToQueue(
        (channel as unknown) as Channel,
        QueueName.TO_PROCESS,
        {
          hfile: {
            file: "file",
            source: "source",
            name: "name",
            dir: "dir",
          },
          root: "root",
          thumbnailsDir: "thumbdir",
        },
        { mandatory: true }
      );

      expect(channel.sendToQueue).toHaveBeenCalled();
      expect(channel.sendToQueue.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe("removeEmptyValues", () => {
    it("should remove empty values", () => {
      expect(
        removeEmptyValues({
          IAmNull: null,
          IAmUndefined: undefined,
          IAmZero: 0,
          IAmEmptyString: "",
        })
      ).toEqual({
        IAmZero: 0,
        IAmEmptyString: "",
      });
    });
  });

  [sortJson, sortJsonStringify].forEach((fct) => {
    describe(fct.name, () => {
      it("should sort the object", () => {
        expect(fct({ zz: 1, aa: 1, bb: 2 })).toMatchSnapshot();
        expect(fct({ zz: { zz: 1, aa: 2 }, aa: 1, bb: 2 })).toMatchSnapshot();
      });

      it("should sort the array", () => {
        expect(fct(["zz", "cc", "bb", "aa"])).toMatchSnapshot();
        expect(fct(["zz", "bb", ["zz", "aa"], "aa"])).toMatchSnapshot();
      });

      it("should sort the mix", () => {
        expect(fct(["zz", { zz: 1, aa: 2 }, "aa"])).toMatchSnapshot();
        expect(
          fct({ zz: 1, aa: ["zz", ["zz", "aa"], "bb", "aa"], bb: 2 })
        ).toMatchSnapshot();
      });
    });
  });

  describe("parentDir", () => {
    it("should return the parent directory", () => {
      expect(parentDir("/root/parent/dir")).toEqual("/root/parent");
      expect(parentDir("/root/parent")).toEqual("/root");
      expect(parentDir("/root")).toEqual("/");
      expect(parentDir()).toEqual("");
    });
  });

  describe("isSupportedMime", () => {
    it("should return a boolean if supported", () => {
      expect(isSupportedMime("image/jpeg")).toEqual(true);
      expect(isSupportedMime("image/png")).toEqual(true);
      // @ts-ignore
      expect(isSupportedMime("doesntexist")).toEqual(false);
    });
  });
});
