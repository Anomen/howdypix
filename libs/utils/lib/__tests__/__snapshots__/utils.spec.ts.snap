// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils:: consume should call the callback with the result of the consume 1`] = `
Array [
  Object {
    "content": "{\\"content\\":{\\"data\\":1}}",
    "data": Object {
      "content": Object {
        "data": 1,
      },
    },
  },
]
`;

exports[`utils:: debugging functions appDebug should call debug() with the right parameters 1`] = `
Array [
  "app:debug:test",
]
`;

exports[`utils:: debugging functions appError should call debug() with the right parameters 1`] = `
Array [
  "app:error:test",
]
`;

exports[`utils:: debugging functions appInfo should call debug() with the right parameters 1`] = `
Array [
  "app:info:test",
]
`;

exports[`utils:: debugging functions appWarning should call debug() with the right parameters 1`] = `
Array [
  "app:warn:test",
]
`;

exports[`utils:: debugging functions libDebug should call debug() with the right parameters 1`] = `
Array [
  "lib:test",
]
`;

exports[`utils:: generateFileUrl should generate the right URLs 1`] = `"http://toto.com/files/@source:dir/file"`;

exports[`utils:: generateFileUrl should generate the right URLs 2`] = `"http://toto.com//files/@source:dir/file"`;

exports[`utils:: generateFileUrl should generate the right paths with missing dir 1`] = `"http://toto.com/files/@source:file"`;

exports[`utils:: generateFileUrl should generate the right paths with missing file 1`] = `"You need to pass a file path to generate a URL."`;

exports[`utils:: generateThumbnailPaths should generate the right paths 1`] = `
Array [
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/dir/filex200.jpg",
    "width": 200,
  },
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/dir/filex600.jpg",
    "width": 600,
  },
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/dir/filex1280.jpg",
    "width": 1280,
  },
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/dir/filex2048.jpg",
    "width": 2048,
  },
]
`;

exports[`utils:: generateThumbnailPaths should generate the right paths with missing dir 1`] = `
Array [
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/filex200.jpg",
    "width": 200,
  },
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/filex600.jpg",
    "width": 600,
  },
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/filex1280.jpg",
    "width": 1280,
  },
  Object {
    "height": null,
    "path": "/thumbnails/.howdypix/source/filex2048.jpg",
    "width": 2048,
  },
]
`;

exports[`utils:: generateThumbnailPaths should generate the right paths with missing file 1`] = `"You must specify a file."`;

exports[`utils:: generateThumbnailUrls should generate the right URLs 1`] = `
Array [
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex200.jpg",
    "width": 200,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex600.jpg",
    "width": 600,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex1280.jpg",
    "width": 1280,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex2048.jpg",
    "width": 2048,
  },
]
`;

exports[`utils:: generateThumbnailUrls should generate the right URLs 2`] = `
Array [
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex200.jpg",
    "width": 200,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex600.jpg",
    "width": 600,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex1280.jpg",
    "width": 1280,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:dir/filex2048.jpg",
    "width": 2048,
  },
]
`;

exports[`utils:: generateThumbnailUrls should generate the right paths with missing dir 1`] = `
Array [
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:filex200.jpg",
    "width": 200,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:filex600.jpg",
    "width": 600,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:filex1280.jpg",
    "width": 1280,
  },
  Object {
    "height": null,
    "url": "http://toto.com/files/@source:filex2048.jpg",
    "width": 2048,
  },
]
`;

exports[`utils:: generateThumbnailUrls should generate the right paths with missing file 1`] = `"You need to pass a file path to generate a URL."`;

exports[`utils:: sendToQueue should call sendToQueue with the right options 1`] = `
Array [
  "toProcess",
  Object {
    "data": Array [
      123,
      34,
      104,
      102,
      105,
      108,
      101,
      34,
      58,
      123,
      34,
      102,
      105,
      108,
      101,
      34,
      58,
      34,
      102,
      105,
      108,
      101,
      34,
      44,
      34,
      115,
      111,
      117,
      114,
      99,
      101,
      34,
      58,
      34,
      115,
      111,
      117,
      114,
      99,
      101,
      34,
      44,
      34,
      110,
      97,
      109,
      101,
      34,
      58,
      34,
      110,
      97,
      109,
      101,
      34,
      44,
      34,
      100,
      105,
      114,
      34,
      58,
      34,
      100,
      105,
      114,
      34,
      125,
      44,
      34,
      114,
      111,
      111,
      116,
      34,
      58,
      34,
      114,
      111,
      111,
      116,
      34,
      44,
      34,
      116,
      104,
      117,
      109,
      98,
      110,
      97,
      105,
      108,
      115,
      68,
      105,
      114,
      34,
      58,
      34,
      116,
      104,
      117,
      109,
      98,
      100,
      105,
      114,
      34,
      125,
    ],
    "type": "Buffer",
  },
  Object {
    "mandatory": true,
  },
]
`;

exports[`utils:: sortJson should sort the array 1`] = `
Array [
  "aa",
  "bb",
  "cc",
  "zz",
]
`;

exports[`utils:: sortJson should sort the array 2`] = `
Array [
  "aa",
  Array [
    "aa",
    "zz",
  ],
  "bb",
  "zz",
]
`;

exports[`utils:: sortJson should sort the mix 1`] = `
Array [
  Object {
    "aa": 2,
    "zz": 1,
  },
  "aa",
  "zz",
]
`;

exports[`utils:: sortJson should sort the mix 2`] = `
Object {
  "aa": Array [
    "aa",
    Array [
      "aa",
      "zz",
    ],
    "bb",
    "zz",
  ],
  "bb": 2,
  "zz": 1,
}
`;

exports[`utils:: sortJson should sort the object 1`] = `
Object {
  "aa": 1,
  "bb": 2,
  "zz": 1,
}
`;

exports[`utils:: sortJson should sort the object 2`] = `
Object {
  "aa": 1,
  "bb": 2,
  "zz": Object {
    "aa": 2,
    "zz": 1,
  },
}
`;

exports[`utils:: sortJsonStringify should sort the array 1`] = `"[\\"aa\\",\\"bb\\",\\"cc\\",\\"zz\\"]"`;

exports[`utils:: sortJsonStringify should sort the array 2`] = `"[\\"aa\\",[\\"aa\\",\\"zz\\"],\\"bb\\",\\"zz\\"]"`;

exports[`utils:: sortJsonStringify should sort the mix 1`] = `"[{\\"aa\\":2,\\"zz\\":1},\\"aa\\",\\"zz\\"]"`;

exports[`utils:: sortJsonStringify should sort the mix 2`] = `"{\\"aa\\":[\\"aa\\",[\\"aa\\",\\"zz\\"],\\"bb\\",\\"zz\\"],\\"bb\\":2,\\"zz\\":1}"`;

exports[`utils:: sortJsonStringify should sort the object 1`] = `"{\\"aa\\":1,\\"bb\\":2,\\"zz\\":1}"`;

exports[`utils:: sortJsonStringify should sort the object 2`] = `"{\\"aa\\":1,\\"bb\\":2,\\"zz\\":{\\"aa\\":2,\\"zz\\":1}}"`;
